chromadb/rate_limiting/admission_controller.py

```diff
@@ -0,0 +1,358 @@
+            if self.waiters[idx] is not None:
+                break
+            self.tail = idx
+
+    def check_invariants(self) -> None:
+        for idx, w in enumerate(self.waiters):
+            if w is None:
+                continue
+            if w.index != idx:
+                raise RuntimeError(
+                    "Admission controller invariants violated: waiter at wrong index"
+                )
+
+        if self.head != self.tail:
+            if self.waiters[self.head] is None:
+                raise RuntimeError(
+                    "Admission controller invariants violated: head is nil"
+                )
+            size = len(self.waiters)
+            tail_idx = (self.tail + size - 1) % size
+            if self.waiters[tail_idx] is None:
+                raise RuntimeError(
+                    "Admission controller invariants violated: tail is nil"
+                )
+
+
+global_admitted = 0
+global_denied = 0
+lock = threading.Lock()
+ac = AdmissionControllerImpl(parallelism=10)  # Number of workers
+
+NUM_REQUESTS = 100
+
+
+def increment_admitted() -> None:
+    global global_admitted
+    with lock:
+        global_admitted += 1
+
+
+def increment_denied() -> None:
+    global global_denied
+    with lock:
+        global_denied += 1
+
+
+def stream_requests(n: int) -> tuple[Optional[Ticket], int]:
+    delay = 0.001  # Seconds per request: 1/N to calculate throughput
+    time.sleep(delay)
+    ticket = ac.admit_one()
+    if ticket:
+        print(f"Admitted request {n}.")
+        return ticket, n
+    else:
+        print(f"Request denied for {n}.")
+        return None, n
+
+
+def process_responses(
+    response_queue: queue.Queue[tuple[Optional[Ticket], int]]
+) -> None:
+    while True:
+        response = response_queue.get()
+        ticket, n = response
+
+        if ticket:
+            time.sleep(0.05)  # Simulate request latency
+            ticket.release()
+            increment_admitted()
+        else:
+            increment_denied()
+
+        if n == NUM_REQUESTS - 1:
+            break
+
+
+def main() -> None:
+    response_queue: queue.Queue[tuple[Optional[Ticket], int]] = queue.Queue()
+    # Number of workers
+    processing_thread = concurrent.futures.ThreadPoolExecutor(max_workers=10).submit(
+        process_responses, response_queue
+    )
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
+        futures = []
+        for i in range(NUM_REQUESTS):
+            future = executor.submit(stream_requests, i)
+            futures.append(future)
+
+        for future in concurrent.futures.as_completed(futures):
+            response = future.result()
+            response_queue.put(response)
+
+    processing_thread.result()
+
+    print(
+        f"Admitted {global_admitted} requests. Denied {global_denied} requests. Total admitted percentage: {global_admitted / NUM_REQUESTS}"
+    )
+
+
+if __name__ == "__main__":
```

Comment:
Is this intended to be a test? Can we write it as one instead of as a main / demo?